server/src/Icepeak/Server/WebsocketServer.hs

```diff
@@ -173,3 +221,67 @@ processUpdates core = go
           go
         -- Stop the loop when we receive a Nothing.
         Nothing -> pure ()
+
+-- * Timeout handling
+--
+-- The websockets library lets you send pings, but it has no built in way to
+-- terminate clients that never send a pong back. We implement this ourselves by
+-- keeping track of when we last received a pong, and then terminating the
+-- connection if the time between the last sent ping and the last received pong
+-- exceeds a certain threshold. See 'WSServerOptions' for more information
+
+-- | The connection-specific on-pong handler. This writes the time at which the
+-- last pong was received so 'pingHandler' can terminate the connection if the
+-- client stops sending pongs back.
+pongHandler :: WSServerOptions -> IO ()
+pongHandler (WSServerOptions lastPongTime) = getTime Monotonic >>= void . swapMVar lastPongTime
+
+-- | An action passed to 'withInterruptiblePingThread' that is used together with
+-- 'pongHandler' to terminate a WebSocket connection if the client stops sending
+-- timely pongs. This returns @True@ if the connection has timed out and should
+-- be terminated.
+pingHandler :: Config -> WSServerOptions -> IO Bool
+pingHandler config (WSServerOptions lastPongTime) = do
+  now <- getTime Monotonic
+  let pingInterval     = TimeSpec (fromIntegral $ configWebSocketPingInterval config) 0
+      pongTimeout      = TimeSpec (fromIntegral $ configWebSocketPongTimeout config) 0
+      lastPongDeadline = now - pingInterval - pongTimeout
+
+  lastPong <- readMVar lastPongTime
+  return $! lastPong < lastPongDeadline
+
+-- | Similar to 'WS.withPingThread', except that it uses a combination of
+-- 'pingHandler' and the 'pongHandler' set in the websocket connection's pong
+-- handler to detect that the thread client stopped sending pongs. If that
+-- happens the @app@ action will be canceled immediately.
+--
+-- The @pingAction@ is exected on every ping, and it should return @True@ if the
+-- client has timed out and the connection should be terminated.
+withInterruptiblePingThread :: WS.Connection -> Int -> IO Bool -> IO () -> IO ()
+withInterruptiblePingThread conn pingInterval pingAction =
+  race_ (interruptiblePingThread conn pingInterval pingAction)
+
+-- | 'WS.pingThread', with the only real difference being that it takes an @IO
+-- Bool@ instead of an @IO ()@ action. If that action returns true, then the
+-- ping thread will return early causing 'withInterruptiblePingThread' to
+-- terminate as well.
+interruptiblePingThread :: WS.Connection -> Int -> IO Bool -> IO ()
+interruptiblePingThread conn pingInterval pingAction
+  | pingInterval <= 0 = return ()
+  | otherwise = ignore `handle` go 1
+  where
+    go :: Int -> IO ()
+    go i = do
+      threadDelay (pingInterval * 1000 * 1000)
+      WS.sendPing conn (T.pack $ show i)
```

Comment:
Couldn't we now theoretically get stuck on this one (due to the lack of timeouts) and thus never reach the place where it is supposed to exit?
Concretely: if the send buffer is already full because the other side is not responding, then this would likely block (though I didn't check) - and the default kernel timeouts for TCP connections are rather long IIRC.

IIRC one of the designs I saw used an external thread with a timer that's reset when receiving pongs, and killing the handler thread when it times out, that would not be prone to this.
(Though it also depends on how `sendPing` is implemented)

Reply:
I'd need to check if it uses a blocking `send()` or not, I also have no idea. A quick glance indicates that it uses this function, which sounds like it would block when the send buffer is full: https://hackage.haskell.org/package/network-3.1.2.9/docs/Network-Socket-ByteString.html

At worst it would behave the same as before, since the `race_` used to spawn the action should prevent the ping thread from outliving the socket as the `app` action will terminate when the socket is closed, causing the other action to be cancelled. But I'll do some digging to figure out what would happen. If the send buffer indeed fills up (though with pings being only a couple bytes long, that would take forever), we could add another `MVar` for the last ping time and then send the pings form a third thread so the current ping handler thread only has to compare the two `MVar`s.

Reply:
I just checked, with a default Ubuntu install it can send `342855` pings before filling the send buffer and blocking. This number will decrease slightly as time goes on (because the ping message becomes larger; it sends a sequence number as a string) but not in a significant way.

This would be fine in practice, but moving the ping sending to another thread and then comparing the values form the `MVar`s in the `interruptiblePingThread` thread does not have this limitation. What do you suggest, go with the third thread just in case or accept that with a 30 second ping interval a 20 year timeout window is probably sufficiently large?

Reply:
I would go with less over-engineering, since we should never run into these limitations.

Reply:
Yes I guess that's true. Adding a third thread would be a lot more robust in that situation.

Reply:
Actually, I think it's not just hypothetical (though Diego might remember that better), but IIRC our initial symptoms of icepeak getting stuck were caused by a send being blocked, which in turn blocked the internal outgoing queue from making progress and eventually it ran full.

Reply:
Yeah, it was the issue, because we didn't decouple it enough. But I wonder if something like that could actually happen here. Maybe going safe is the way then.

Reply:
I thought about this for some more and just adding a third thread may not be entirely correct since it would need to use its own delay/timer routine, which may slowly drift out of phase compared to the thread that actually sends the pings. Using an actual timer like you suggested would work, but we don't already depend on a package that can do this like [`timers`](https://hackage.haskell.org/package/timers-0.2.0.4/docs/Control-Concurrent-Timer.html). I checked the implementation and I don't think it's that much less expensive to restart a timer there than it would be to just slap a `timeout` onto the `WS.sendPing`, which works as expected. Any objections to just doing that instead? That prevents the ping from blocking indefinitely when the send buffer is full, and it interacts nicely with the existing timeout checking function.

Reply:
> Doesn't the current implementation also run out of phase, because the ping thread runs in intervals of "ping interval" while the pongs are expected after the "pong timeout"?

There's a bit of play between `pingAction` being evaluated and the on-pong handler being evaluated but that should follow a normal distribution and not really increase or decrease over time. The phase drifting I was talking about would happen if you used another thread for checking the `IORef TimeSpec`s written by the ping- and pong handler instead of using an actual resettable timer. But I think the `timeout` here is an even nicer solution since it doesn't need to introduce any more moving parts.

> Less expensive in terms of what? timeout also uses a timer internally. It's certainly less complex though - but what would be the timeout for that operation then?

I set the timeout to the ping interval. Since if the socket blocks for that long, then the `pingHandler` will also consider the connection to have timed out so it works out nicely without any other changes.

Reply:
Sounds like a plan!